### DISCUSSION TITLE
Locking Durable Entities from within client functions

### DISCUSSION THREAD

**Discussion by a3y3** [Azure/azure-functions-durable-extension#1938] [Q&A] [answered]
I understand what `lockasync` does and how it behaves, but I can't understand why this feature is restricted to only orchestrator functions. Is there any way to use Durable Locks from within say, a client function? (I can't convert my client function to an orchestrator function; it's not deterministic)

**Comment by cgillum** [comment, marked as answer]
Lock operations on entities involve two messages: one that is sent to the entity to lock it and one that is sent back to the sender to confirm that the lock was acquired. A lock can be held for an indefinite amount of time and also has to keep track of who owns the lock. Orchestrations are compatible with all these characteristics since they are long-running and have permanent identifiers.

Client functions, on the other hand, are transient and not capable of waiting indefinitely for an ACK. They also don't have any kind of permanent identifier for the entity to use when deciding whether to allow a caller to perform operations. I suppose a non-orchestrator client could come up with an identifier and send it as part of the lock request, but I think there's still a problem of notifying a transient client when a lock has been successfully acquired. There's also the question of how you ensure that the lock gets released reliably.

FYI @sebastianburckhardt in case I got anything wrong, or if there are other important explanations.

**Reply by a3y3** [reply]
I think that makes sense to me, thanks! I wish I could mark it as an answer but Github is throwing 422s whenever I click that button.

**Reply by a3y3** [reply]
@cgillum If I made my own durable lock using Durable Entities (create a Lock class and have fields that indicate the status of the lock) would that be a worthwhile effort? Just checking if there aren't any obvious pitfalls to this approach that I'm missing before I start to implement something like that.

**Reply by cgillum** [reply]
What would be the reason for creating your own implementation? Is it because you want to be able to use it from a transient client function? Whether it would be worthwhile probably depends on your scenario - i.e. how you intend to use it and what your constraints are.
